Upgrade JavaFX to ver 17

Eclipse-Dominator opened the pull request:
Fixes #193 
Fixes #108 

This is an alternative PR to #196 which explores adding an extra log message to inform the user to ignore the warning message while upgrading to an LTS version.

The application uses JavaFX 11 which is reaching its EOL in September
2023. Thus it is better to upgrade JavaFX to the next LTS version which is version 17.

However, after version 16, JavaFX now throws a warning when JavaFX modules are loaded from the classpath. As there is currently no plan to migrate to using Modules in AB3, let's include an additional log message to inform the users to ignore the warning message.

<img width="671" alt="image" src="https://github.com/se-edu/addressbook-level3/assets/4567895/fca77461-665b-4c7c-a316-576a8dcc5d9d">
<img width="683" alt="image" src="https://github.com/se-edu/addressbook-level3/assets/4567895/cdb75d73-034b-4078-a05b-f6bd18f52750">

damithc commented:
@Eclipse-Dominator also post a screenshot of how the log output looks.

damithc commented on src/main/java/seedu/address/Main.java:
```diff
@@ -19,7 +22,23 @@
+
+    /**
+     * Main entry point to the application.
+     */
```
Seems redundant. Remove?

Eclipse-Dominator replied:
Removing this seems to cause Checkstyle to throw missing javadoc warning

damithc replied:
I see. But why now? The method was there to begin with, right?

damithc commented on src/main/java/seedu/address/Main.java:
```diff
@@ -19,7 +22,23 @@
+         * This warning however, can be safely ignored. Thus, this log is used to inform the user that the warning
+         * can be ignored.
+         * https://github.com/openjdk/jfx/blob/master/doc-files/release-notes-16.md
+         */
```
Original comment was OK. But rewrote to see if the flow can be refined further.
```suggestion
        /*
         * As per https://github.com/openjdk/jfx/blob/master/doc-files/release-notes-16.md
         * JavaFX 16 (or later) runtime logs a warning at startup if JavaFX classes are loaded from
         * the classpath instead of a module.
         * Our application does not use Java modules yet. Even if it did, modules are ignored when
         * packed into a FAT Jar file (as we do), which means this warning will persist even then.
         * The warning however, can be safely ignored. Thus, the following log informs
         * the user (if looking at the log output) that the said warning appearing in the log 
         * can be ignored.
         */
```

damithc reviewed (changes requested):
Just a couple of cosmetic suggestions.

Eclipse-Dominator committed:
Exempt main method from javadoc requirement

Let's update checkstyle.xml to omit the main method from
the javadoc requirement, as the purpose of the main method
is clear.

damithc reviewed (commented):
Looks good.

Suggested commit message (shorter):
```
Exempt main method from javadoc requirement

Let's update checkstyle.xml to omit the main method from
the javadoc requirement, as the purpose of the main method
is clear.
```

Minor, the checkstyle commit should come before the JavaFX commit?

damithc commented:
@se-edu/tech-team-level1 for your review ...

damithc commented on src/main/java/seedu/address/Main.java:
```diff
@@ -19,7 +22,20 @@
+         * The warning however, can be safely ignored. Thus, the following log informs
+         * the user (if looking at the log output) that the said warning appearing in the log
+         * can be ignored.
+         */
```
Should this be a `//` comment instead?

chia-yh reviewed (commented):
LGTM, however, should the following portion of the `.fxml` files (`HelpWindow.fxml`, etc.) be updated as well?
`xmlns="http://javafx.com/javafx/11"`

dlimyy reviewed (commented):
LGTM! Similarly, as mentioned by @chia-yh, I believe that all the `.fxml` should be updated to reflect the latest javafx version.

lywich reviewed (commented):
LGTM too! I concur with @chia-yh that it would be prudent to update the `.fxml` files to the latest version.

markusyeo commented:
LGTM!

In addition to @chia-yh 's catch

I did a quick search of other instances of `http://javafx.com/javafx/` appearing in the fxml files and noticed that there are other namespaces still pointing at `http://javafx.com/javafx/8`.

I think we might need to look at updating all of them to be prudent, or consider removing the version too

jundatan reviewed (commented):
LGTM. Other than the updating of  `.fxml` which has already been mentioned by other members.

SPWwj reviewed (changes requested):
There's a warning that need to resolve.
<img width="1004" alt="image" src="https://github.com/se-edu/addressbook-level3/assets/30100720/ef4d9870-b556-4a01-8c59-1e1a10212624">

Additionally, all .fxml files in `resources\view` should be updated to the latest version.

Eclipse-Dominator commented:
> There's a warning that need to resolve. <img alt="image" width="1004" src="https://user-images.githubusercontent.com/30100720/252539612-ef4d9870-b556-4a01-8c59-1e1a10212624.png">
> 
> Additionally, all .fxml files in `resources\view` should be updated to the latest version.

I included a separate info log to tell the users to discard this warning. It seems intellij filters away info logs so I updated the logger to use warning instead and it will appear in intellij as such:

![image](https://github.com/se-edu/addressbook-level3/assets/4567895/c6b43fd4-67bd-47fe-9efb-80141866d936)

chia-yh reviewed (commented):
LGTM

SPWwj commented:
> I included a separate info log to tell the users to discard this warning. It seems intellij filters away info logs so I updated the logger to use warning instead and it will appear in intellij as such:

I'm unsure if we can resolve this warning. This solution look like a last resort.

hansstanley commented:
Small issue since it's a test file, but for consistency should the `xmlns` and `xmlns:fx` of [`validFileWithFxRoot.fxml`](https://github.com/Eclipse-Dominator/addressbook-level3/blob/master/src/test/resources/view/UiPartTest/validFileWithFxRoot.fxml) be updated too? Currently it does not specify version:

```fxml
<fx:root type="seedu.address.ui.TestFxmlObject" xmlns="http://javafx.com/javafx"
            xmlns:fx="http://javafx.com/fxml">
    <text>Hello World!</text>
</fx:root>
```

Eclipse-Dominator commented:
> > I included a separate info log to tell the users to discard this warning. It seems intellij filters away info logs so I updated the logger to use warning instead and it will appear in intellij as such:
> 
> I'm unsure if we can resolve this warning. This solution look like a last resort.

The proper solution to resolve the issue is not very clean. It mainly involves migrating to use java modules which incur the following problems:
- Upgrading of `jackson` to a later version (in order to use modules) which leads to changes in behavior of the code when converting `Path` objects to file
- The compiled fat Jar file will ignore the modules so the warning will still show up in the final product even though it might not show up after moving to modules. (currently I have not found a way to prevent this)

Alternatively, we can also manually suppress the warning by changing the JavaFX logger but it might not be ideal since using modules is the supposedly "correct" way of using JavaFX.

damithc commented:
> Small issue since it's a test file, but for consistency should the `xmlns` and `xmlns:fx` of [`validFileWithFxRoot.fxml`](https://github.com/Eclipse-Dominator/addressbook-level3/blob/master/src/test/resources/view/UiPartTest/validFileWithFxRoot.fxml) be updated too? Currently it does not specify version:
> 
> ```
> <fx:root type="seedu.address.ui.TestFxmlObject" xmlns="http://javafx.com/javafx"
>             xmlns:fx="http://javafx.com/fxml">
>     <text>Hello World!</text>
> </fx:root>
> ```

@Eclipse-Dominator any response to this comment?

SPWwj commented:
> > > I included a separate info log to tell the users to discard this warning. It seems intellij filters away info logs so I updated the logger to use warning instead and it will appear in intellij as such:

The warnings are issued by the JVM and are emitted before the Java code even starts to run. I think your current solution is effectively handle the situation, as the warning doesn't affect the application.

Eclipse-Dominator committed:
Upgrade JavaFX to ver 17

The application uses JavaFX 11 which is reaching its EOL in September
2023. Thus it is better to upgrade JavaFX to the next LTS version which
is version 17.

However, after version 16, JavaFX now throws a warning when JavaFX
modules are loaded from the classpath. As there is currently no plan to
migrate to using Modules in AB3, let's include an additional log message
to inform the users to ignore the warning message.

damithc commented:
> @Eclipse-Dominator any response to this comment?

@Eclipse-Dominator reminder ...

Eclipse-Dominator commented:
> > Small issue since it's a test file, but for consistency should the `xmlns` and `xmlns:fx` of [`validFileWithFxRoot.fxml`](https://github.com/Eclipse-Dominator/addressbook-level3/blob/master/src/test/resources/view/UiPartTest/validFileWithFxRoot.fxml) be updated too? Currently it does not specify version:
> > ```
> > <fx:root type="seedu.address.ui.TestFxmlObject" xmlns="http://javafx.com/javafx"
> >             xmlns:fx="http://javafx.com/fxml">
> >     <text>Hello World!</text>
> > </fx:root>
> > ```
> 
> @Eclipse-Dominator any response to this comment?

I have updated the format `xmlns` and `xmlns:fx` to follow the current project versions.

damithc commented:
> I have updated the format `xmlns` and `xmlns:fx` to follow the current project versions.

Did you post a new version (via CIHR) after this change?

Eclipse-Dominator commented:
> > I have updated the format `xmlns` and `xmlns:fx` to follow the current project versions.
> 
> Did you post a new version (via CIHR) after this change?

I encountered some issues with CIHR. It appears that
> PR not up to date with base (master) head. Rebase required.

even though the current `upgrade-to-jfx-17` branch is already update to date with master

damithc commented:
> even though the current `upgrade-to-jfx-17` branch is already update to date with master

@Eclipse-Dominator How did you update it? I suspect CIHR requires the branch to be synced with master via a rebase, not a merge.

damithc commented:
@Eclipse-Dominator Also, update the master branch of your fork (in case that matters). It seems to be behind the upstream master.

Eclipse-Dominator commented:
> > even though the current `upgrade-to-jfx-17` branch is already update to date with master
> 
> @Eclipse-Dominator How did you update it? I suspect CIHR requires the branch to be synced with master via a rebase, not a merge.

@damithc It's done via a rebase and I can confirm that the current branch is up to date with upstream's master branch. 
I will try to do some force pushes to see if this issue can be resolved.

![image](https://github.com/se-edu/addressbook-level3/assets/4567895/fbe39f83-37fa-40a4-8e29-b0584e3d26f8)
